libraries/reconstruction_cuda/cuda_gpu_iirconvolve.cu

```diff
@@ -1,107 +1,314 @@
-// inspired by https://devblogs.nvidia.com/efficient-matrix-transpose-cuda-cc/
-// TILE_DIM=32, BLOCK_ROWS=8
-// No bank-conflict transpose
-// Same as transposeCoalesced except the first tile dimension is padded
-// to avoid shared memory bank conflicts.
-// can be used to transpose non-square 2D arrays
-__global__
-void transposeNoBankConflicts32x8(float *odata, const float *idata, int xdim, int ydim) {
-    __shared__ float tile[32][32 + 1];
-    int tilex = blockIdx.x * 32;
-    int tiley = blockIdx.y * 32;
-    int x = tilex + threadIdx.x;
-    int y = tiley + threadIdx.y;
-
-    for (int j = 0; j < 32; j += 8) {
-        int index = (y + j) * xdim + x;
-        if (index < (xdim*ydim)) {
-            tile[threadIdx.y + j][threadIdx.x] = idata[index];
+#include "cuda_utils.h"
+
+namespace iirConvolve2D_Cardinal_BSpline_3_MirrorOffBoundKernels {
+
+const int WARP_SIZE = 32;
+
+const int BLOCK_SIZE = 64; // { 16, 32, 64 }
+const int COL_BLOCK_SIZE = 128; // { 32, 64, 128, 256, 512, 1024} ... above 128 only if it is multiple of dim size
+const int UNROLL = 8;
+
+using shared_type = float[BLOCK_SIZE][BLOCK_SIZE+1];
+using data_ptr = float * __restrict__;
+
+__global__ void sum_lines(data_ptr in, int x) {
+    const float z = sqrtf(3.f) - 2.f;
+    const float z_0 = (1.f + z) / z;
+
+    int line = blockIdx.x * x;
+    int tid = threadIdx.x;
+
+    __shared__ volatile float sdata[WARP_SIZE];
+    if (tid == 0) {
+         sdata[tid] = in[line + tid] * z_0;
+    }
+    else {
+        sdata[tid] = in[line + tid] * powf(z, tid);
+    }
+
+    if (tid < 16) sdata[tid] += sdata[tid + 16];
+    if (tid < 8)  sdata[tid] += sdata[tid + 8];
+    if (tid < 4)  sdata[tid] += sdata[tid + 4];
+    if (tid < 2)  sdata[tid] += sdata[tid + 2];
+
+    if (tid == 0) {
+        in[line] = sdata[0] + sdata[1];
+    }
+}
+
+__global__ void sum_cols(data_ptr in, int x) {
+    const float z = sqrtf(3.f) - 2.f;
+    const float z_0 = (1.f + z) / z;
+
+    int col = blockIdx.x;
+    int tid = threadIdx.x;
+
+    __shared__ volatile float sdata[WARP_SIZE];
+    if (tid == 0) {
+         sdata[tid] = in[col + tid * x] * z_0;
```

Comment:
except for this indexing, looks exactly the same as sum_lines. Is it possible to unify the implementation?